statsmodels/base/_penalties.py

```diff
@@ -73,54 +86,457 @@ def grad(self, params):
+
+
+class NonePenalty(Penalty):
+    """
+    A penalty that does not penalize.
+    """
+
+    def __init__(self, **kwds):
+        if kwds:
+            import warnings
+            warnings.warn('keywords will be ignored')
+
+    def func(self, params):
+        if params.ndim == 2:
+            return np.zeros(params.shape[1:])
+        else:
+            return 0
+
+    def deriv(self, params):
+        return np.zeros(params.shape)
+
+    def hessian(self, params):
+        # returns diagonal of hessian
+        return np.zeros(params.shape[0])
+
 
 class L2(Penalty):
     """
     The L2 (ridge) penalty.
     """
 
-    def __init__(self, wts=None):
-        if wts is None:
-            self.wts = 1.
+    def __init__(self, weights=None, wts=None):
+        weights = _check_wts(weights, wts)  # for deprecation wts
+        if weights  is None:
+            self.weights  = 1.
         else:
-            self.wts = wts
+            self.weights  = weights
         self.alpha = 1.
 
     def func(self, params):
-        return np.sum(self.wts * self.alpha * params**2)
+        return np.sum(self.weights * self.alpha * params**2)
 
-    def grad(self, params):
-        return 2 * self.wts * self.alpha * params
+    def deriv(self, params):
+        return 2 * self.weights * self.alpha * params
+
+    def deriv2(self, params):
+        return 2 * self.weights * self.alpha * np.ones(len(params))
+
+
+class L2Univariate(Penalty):
+    """
+    The L2 (ridge) penalty applied to each parameter.
+    """
+
+    def __init__(self, weights=None):
+        if weights is None:
+            self.weights = 1.
+        else:
+            self.weights = weights
+
+    def func(self, params):
+        return self.weights * params**2
+
+    def deriv(self, params):
+        return 2 * self.weights * params
+
+    def deriv2(self, params):
+        return 2 * self.weights * np.ones(len(params))
 
 
 class PseudoHuber(Penalty):
     """
     The pseudo-Huber penalty.
     """
 
-    def __init__(self, dlt, wts=None):
+    def __init__(self, dlt, weights=None, wts=None):
         self.dlt = dlt
-        if wts is None:
-            self.wts = 1.
+        weights = _check_wts(weights, wts)  # for deprecation wts
+        if weights  is None:
+            self.weights  = 1.
         else:
-            self.wts = wts
+            self.weights  = weights
         self.alpha = 1.
 
     def func(self, params):
         v = np.sqrt(1 + (params / self.dlt)**2)
         v -= 1
         v *= self.dlt**2
-        return np.sum(self.wts * self.alpha * v)
+        return np.sum(self.weights * self.alpha * v, 0)
 
-    def grad(self, params):
+    def deriv(self, params):
         v = np.sqrt(1 + (params / self.dlt)**2)
-        return params * self.wts * self.alpha / v
+        return params * self.weights * self.alpha / v
+
+    def deriv2(self, params):
+        v = np.power(1 + (params / self.dlt)**2, -3/2)
+        return self.weights * self.alpha * v
```

Comment:
Is there a term missing here?  It's a bit tough b/c `v` is slightly different between here vs `deriv` vs `func`, but eyeballing it this looks like a `deriv / params` term might be missing.

Reply:
It's correct the way it is, verified by the unit tests.
After a few failed attempts to derive it by hand, I used wolframalpha, and this is a simplification of the expression there.
v had a different expression before, but I kept the same name for the temp variable